pixl_core/src/core/upload.py

```diff
@@ -11,179 +11,149 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-"""Functionality to upload files to an endpoint."""
+"""Functionality to upload files to a remote server."""
 
 from __future__ import annotations
 
 import ftplib
 import logging
-import ssl
+from abc import ABC, abstractmethod
 from datetime import datetime, timezone
-from ftplib import FTP_TLS
 from pathlib import Path
-from typing import TYPE_CHECKING, Any, BinaryIO
+from typing import TYPE_CHECKING, BinaryIO
 
 from decouple import config
 
 if TYPE_CHECKING:
-    from socket import socket
-
     from core.exports import ParquetExport
+    from core.project_config import PixlConfig
 
 
-from core.db.queries import get_project_slug_from_db, update_exported_at
+from core._secrets import AzureKeyVault
+from core._upload_ftps import (
+    _connect_to_ftp,
+    _create_and_set_as_cwd,
+    _create_and_set_as_cwd_multi_path,
+)
+from core.db.queries import get_project_slug_from_hashid, update_exported_at
 
 logger = logging.getLogger(__name__)
 
 
-class ImplicitFtpTls(ftplib.FTP_TLS):
-    """
-    FTP_TLS subclass that automatically wraps sockets in SSL to support implicit FTPS.
+class Uploader(ABC):
+    """Upload strategy interface."""
+
+    @abstractmethod
+    def __init__(self, project_config: PixlConfig) -> None:
+        """
+        Initialise the uploader for a specific project with the destination configuration and an
+        AzureKeyvault instance. The keyvault is used to fetch the secrets required to connect to
+        the remote destination.
+
+        :param project: The project name for which the uploader is being initialised. Used to fetch
+            the correct secrets from the keyvault.
+        """
+        self.project_config = project_config
+        self.keyvault = AzureKeyVault()
```

Comment:
If we're going to make an abstract base class then why don't we have methods like _set_config which you have to override, and upload_dicom etc. Like we'll always want to run _set_config on initialisation right?

You could even have a static method in this class which takes in the project config and creates the right child class and returns it. That way there's a single point for parsing the config value for both dicom and parquet?

Reply:
Yeah I didn't want to put too much in the base class because I wasn't sure how similar/different the other upload strategies would be for Azure and DICOMweb as we don't really have a plan for those yet AFAIK?
But the `_set_config` makes sense. And maybe `upload_dicom` as well, but `upload_parquet_files` would only exist for some upload strategies.

Reply:
yeah can always have upload partquet files return raise a not implemented error for those types?